### PR TITLE
Add steps to createOffer and explicitly specify what is queued

### LINKED ISSUE #744: createOffer should have steps that explicitly states what is queued

### PR DESCRIPTION
**Pull request by adam-be**
Fixes #744 

This PR is currently based on PR #721 to make use of some changes to the generic queued operation section.

### CODE REVIEW

**Comment by adam-be** [comment]
Rebased passed PR #721 so this PR should be easier to review now. @jan-ivar, can you please have a look?

File: webrtc.html [hunk 1]
```diff
@@ -1251,30 +1251,99 @@
...
+                                <p>If <var>connection</var>'s [[<a>isClosed</a>]]
+                                slot is <code>true</code>, then abort these
+                                steps.</p>
+                              </li>
+                              <li>
+                                <p>If the identity provider was unable to
+                                produce an identity assertion, reject
+                                <var>p</var> with a <code>DOMException</code>
+                                object whose <code>name</code> attribute has
+                                the value <code>NotReadableError</code> and
```
Comment (jan-ivar) [inline webrtc.html]: comma

File: webrtc.html [hunk 2]
```diff
@@ -1251,30 +1251,99 @@
...
+                                abort these steps.</p>
+                              </li>
+                              <li>
+                                <p>Reject <var>p</var> with a
+                                <code>DOMException</code> object whose
+                                <code>name</code> attribute has the value
+                                <code>OperationError</code> and abort these
```
Comment (jan-ivar) [inline webrtc.html]: comma

File: webrtc.html [hunk 3]
```diff
@@ -1251,30 +1251,99 @@
...
+                      </li>
+                      <li>
+                        <p>In parallel, start the process to generate an SDP
+                        offer, as described in <span data-jsep=
+                        "createoffer">[[!JSEP]]</span>.</p>
+                        <ol>
+                          <li>
+                            <p>If the process to generate an SDP offer failed
+                            for any reason, the User Agent MUST queue a task
+                            that runs the following steps:</p>
```
Comment (jan-ivar) [inline webrtc.html]: JSEP doesn't cover identity AFAIK, so logically does this need to say "If the process to generate an SDP offer failed, or if the identity provider was unable to produce an identity assertion, then ..." ?

**Comment by jan-ivar** [comment]
Lgtm once `isClosed` check is moved out.

File: webrtc.html [hunk 4]
```diff
@@ -1251,30 +1251,99 @@
...
+                  </li>
+                  <li>
+                    <p>Return the result of <a href=
+                    "#enqueue-an-operation">enqueuing</a> the following
+                    steps:</p>
+                    <ol>
+                      <li>
+                        <p>If <var>connection</var>'s [[<a>isClosed</a>]] slot
+                        is <code>true</code>, return a promise rejected with an
+                        <code>InvalidStateError</code>.</p>
```
Comment (jan-ivar) [inline webrtc.html]: I mean we suppress rejections from *enqueued* actions past close.

Comment (adam-be) [inline webrtc.html]: This text is in line with the queued steps of addIceCandidate, but I think what you're saying makes sense. So we check for closed state right away, and then make sure that nothing is taken of the queue if the isClosed slot is true.

**Comment by adam-be** [comment]
I've moved the rejection of closed state into the queueing steps. I've also added checks into the queuing algorithm that checks if the isClosed flag is set before advancing (e.g. taking the next operation off the queue). @jan-ivar, please review (especially the third commit).

File: webrtc.html [hunk 5]
```diff
@@ -1300,14 +1302,14 @@
...
-                                the value <code>NotReadableError</code> and
+                                the value <code>NotReadableError</code>, and
                                 abort these steps.</p>
                               </li>
                               <li>
                                 <p>Reject <var>p</var> with a
                                 <code>DOMException</code> object whose
                                 <code>name</code> attribute has the value
-                                <code>OperationError</code> and abort these
+                                <code>OperationError</code> and, abort these
```
Comment (jan-ivar) [inline webrtc.html]: Comma in wrong place, though I just noticed this isn't an if-statement, and there are no more steps, so can we simply remove ", and abort these steps?"

File: webrtc.html [hunk 6]
```diff
@@ -1251,30 +1280,96 @@
...
+                                <var>p</var> with a <code>DOMException</code>
+                                object whose <code>name</code> attribute has
+                                the value <code>NotReadableError</code>, and
+                                abort these steps.</p>
+                              </li>
+                              <li>
+                                <p>Reject <var>p</var> with a
+                                <code>DOMException</code> object whose
+                                <code>name</code> attribute has the value
+                                <code>OperationError</code> and, abort these
```
Comment (jan-ivar) [inline webrtc.html]: Comma in wrong place, though I just noticed this isn't an if-statement, and there are no more steps, so can we simply remove ", and abort these steps?"

File: webrtc.html [hunk 7]
```diff
@@ -574,6 +574,15 @@
         following steps:</p>
         <ol>
           <li>
+            <p>Let <var>connection</var> be the current <code>
+            <a>RTCPeerConnection</a></code> object.</p>
+          </li>
+          <li>
+            <p>If <var>connection</var>'s [[<a>isClosed</a>]] slot is
+            <code>true</code>, return a promise rejected with an
+            <code>InvalidStateError</code>.</p>
```
Comment (jan-ivar) [inline webrtc.html]: I notice we're inconsistent in this spec in requiring ", and abort these steps" after `return` and `throw` statements. Do we want to clean that up separately?

File: webrtc.html [hunk 8]
```diff
@@ -588,17 +597,37 @@
...
-            execute the following steps:</p>
+            <var>operation</var>, run the following steps:</p>
             <ol>
               <li>
-                <p>Remove the first element of [[<a>operations</a>]].</p>
+                <p>If <var>connection</var>'s [[<a>isClosed</a>]] slot is
+                <code>true</code>, abort these steps.</p>
+              </li>
+              <li>
+                <p>If the promise returned by <var>operation</var> fulfilled
```
Comment (jan-ivar) [inline webrtc.html]: was fulfilled

File: webrtc.html [hunk 9]
```diff
@@ -588,17 +597,37 @@
...
+                with a value, fulfill <var>p</var> with that value.</p>
+              </li>
+              <li>
+                <p>If the promise returned by <var>operation</var> rejected
```
Comment (jan-ivar) [inline webrtc.html]: was rejected

**Comment by jan-ivar** [comment]
Lgtm with nits.

File: webrtc.html [hunk 10]
```diff
@@ -1251,30 +1280,96 @@
...
+                                <p>If <var>connection</var>'s [[<a>isClosed</a>]]
+                                slot is <code>true</code>, then abort these
+                                steps.</p>
+                              </li>
+                              <li>
+                                <p>If the identity provider was unable to
+                                produce an identity assertion, reject
+                                <var>p</var> with a <code>DOMException</code>
+                                object whose <code>name</code> attribute has
+                                the value <code>NotReadableError</code>, and
```
Comment (fluffy) [inline webrtc.html]: It would be nice to get more information about what went wrong

Comment (adam-be) [inline webrtc.html]: Yeah. We might want to create a separate issue on that.

**Comment by adam-be** [comment]
Thanks @jan-ivar and @fluffy for reviewing

**Comment by adam-be** [comment]
Rebased on top of commit that patches the JSEP reference system. Also updated a JSEP reference that previously wasn't working. I think this is ready to be merged now.

**Pull request merged**